vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java

```diff
@@ -0,0 +1,188 @@
+                abStack.get(abStack.size() - 1).reset(abvsStack.get(abvsStack.size() - 1));
+            }
+            if (token == JsonToken.START_OBJECT) {
+                if (!nested) {
+                    abvsStack.add(new ArrayBackedValueStorage());
+                    obStack.add(new ObjectBuilder());
+                }
+                itemStack.add("object");
+                lastItem = "object";
+                abvsStack.get(abvsStack.size() - 1).reset();
+                obStack.get(obStack.size() - 1).reset(abvsStack.get(abvsStack.size() - 1));
+            }
+            if (token == JsonToken.FIELD_NAME) {
+                UTF8StringPointable sp = new UTF8StringPointable();
+                ArrayBackedValueStorage key = new ArrayBackedValueStorage();
+                key.reset();
+                spStack.add(sp);
+                DataOutput outk = key.getDataOutput();
+                svb.write(parser.getText(), outk);
+                spStack.get(spStack.size() - 1).set(key);
+            }
+            if (token == JsonToken.VALUE_NUMBER_INT) {
+                atomic.reset();
+                out.write(ValueTag.XS_INTEGER_TAG);
+                out.writeLong(parser.getLongValue());
+                if (lastItem == "array") {
+                    abStack.get(abStack.size() - 1).addItem(atomic);
+                } else if (lastItem == "object") {
+                    obStack.get(obStack.size() - 1).addItem(spStack.get(spStack.size() - 1), atomic);
+                }
+            }
+            if (token == JsonToken.VALUE_STRING) {
+                atomic.reset();
+                out.write(ValueTag.XS_STRING_TAG);
+                svb.write(parser.getText(), out);
+                if (lastItem == "array") {
+                    abStack.get(abStack.size() - 1).addItem(atomic);
+                } else if (lastItem == "object") {
+                    obStack.get(obStack.size() - 1).addItem(spStack.get(spStack.size() - 1), atomic);
+                }
+            }
+            if (token == JsonToken.VALUE_NUMBER_FLOAT) {
+                atomic.reset();
+                out.write(ValueTag.XS_DOUBLE_TAG);
+                out.writeDouble(parser.getDoubleValue());
+                if (lastItem == "array") {
+                    abStack.get(abStack.size() - 1).addItem(atomic);
+                } else if (lastItem == "object") {
+                    obStack.get(obStack.size() - 1).addItem(spStack.get(spStack.size() - 1), atomic);
+                }
+            }
+            if (token == JsonToken.END_ARRAY) {
+                if (itemStack.size() == 1) {
+                    abStack.get(abStack.size() - 1).finish();
+                    abStack.remove(abStack.size() - 1);
+                    itemStack.remove(itemStack.size() - 1);
+                    nested = true;
+                    lastItem = "array";
+                } else if (itemStack.get(itemStack.size() - 2) == "array") {
```

Comment:
Comparing using == compares for object identity, not for string equality. It works in this case, as the strings are literals in the java source code and thus the same string will be represented by the same object. However, this is risky code. Using an enum instead of a string solves this problem as well.